### PR TITLE
:sparkles: Starting aws registration by spoke by assuming IAM role on startup and adding annotations to ManagedCluster CR

### PR DESCRIPTION
**Pull request by jaswalkiranavtar**
## Summary
This PR contains
- AWS role assumption on start up by spoke
- Adding annotations to ManagedCluster CR so that hub can recognize it as a registration request with AWS IAM auth
- Mounting dot-aws directory needed by aws-cli

## Related issue(s)
https://github.com/open-cluster-management-io/ocm/issues/514

### CODE REVIEW

**Review by mikeshng — COMMENTED** [review]
This is great! Thank you for your contribution!

Left some comments. Regarding showing the error on klusterlet condition, maybe wait for the operator maintainers and/or Qiu Jian's feedback.

File: manifests/klusterlet/managed/klusterlet-registration-serviceaccount.yaml
Comment (jaswalkiranavtar) [reply manifests/klusterlet/managed/klusterlet-registration-serviceaccount.yaml]: addressed

File: manifests/klusterlet/managed/klusterlet-registration-serviceaccount.yaml
Comment (jaswalkiranavtar) [reply manifests/klusterlet/managed/klusterlet-registration-serviceaccount.yaml]: addressed

File: manifests/klusterlet/managed/klusterlet-work-serviceaccount.yaml
Comment (jaswalkiranavtar) [reply manifests/klusterlet/managed/klusterlet-work-serviceaccount.yaml]: addressed

File: manifests/klusterlet/managed/klusterlet-work-serviceaccount.yaml
Comment (jaswalkiranavtar) [reply manifests/klusterlet/managed/klusterlet-work-serviceaccount.yaml]: addressed

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: removed new library

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: removed new library

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (dongbeiqing91) [inline pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:463]: This function is not robust enough. Although the format of hubClusterArn and managedClusterArn could have been guaranteed somewhere else, this function should still validate the format of these ARNs in the klusterlet to prevent potential runtime errors.

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (dongbeiqing91) [inline pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:585]:
```suggestion
	hash := md5.Sum([]byte(strings.Join([]string{hubClusterAccountId, hubClusterName, managedClusterAccountId, managedClusterName}, "#"))) // #nosec G401
```

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:463]: Thr validation is done in the same file just before invoking this function. If the validation fails, this function will not be invoked. Is that not sufficient?

Comment (dongbeiqing91) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:463]: why not moving the validation into the function?

File: test/integration/operator/klusterlet_singleton_test.go
Comment (jaswalkiranavtar) [reply test/integration/operator/klusterlet_singleton_test.go:192]: The expression returns false if the annotation.is missing but i get your point that we dont have to do comparison, will fix it.

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:345]: Please share a sample where such a condition will be generated.

Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:345]: Good idea validating in crd, lemme.give it a try.

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:470]: Sure, how about a type called ManagedClusterIamRole having arn() and md5HashSuffix() functions?

Comment (qiujian16) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:470]: sounds good

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (qiujian16) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:345]: https://github.com/open-cluster-management-io/ocm/blob/main/pkg/operator/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go#L201C2-L201C22

Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:345]: @qiujian16 created a PR in api repo:

https://github.com/open-cluster-management-io/api/pull/347

File: test/integration/operator/klusterlet_singleton_test.go
Comment (jaswalkiranavtar) [reply test/integration/operator/klusterlet_singleton_test.go]: fixed

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: fixed

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: fixed

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: fixed

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (jaswalkiranavtar) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: not needed as the validation is moved to CRD

**Comment by jaswalkiranavtar** [comment]
Thanks for approving api repo PR. Do you have any further comments in this one?

File: pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go
Comment (mikeshng) [reply pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: Verified that it has been removed.

File: test/integration/operator/klusterlet_singleton_test.go
Comment (mikeshng) [reply test/integration/operator/klusterlet_singleton_test.go]: Verified that it has been addressed.

**Review by mikeshng — APPROVED** [review]
Great work on this PR!

Huge thank you to all the code reviewers for taking the time to provide valuable feedback. I’ve ensured that all your comments have been addressed. If there are any additional concerns, @jaswalkiranavtar and his team will address them in another PR quickly.

Typically, I am not the one to approve PRs involving this area of the codebase. However, I’m making an exception here as, based on my review, all changes are contained within the hub on EKS associated flags. This containment minimizes the risk of regression to the main/latest branch. To accommodate the contributors' schedules and planning, I’m approving this merge based on the prior reviews, comments, and my own assessment.

/approve

/lgtm

**Review by jnpacker — APPROVED** [review]
Reviewed the code and comments. /LGTM

**Pull request merged**
